source/common/http/http_server_properties_cache_impl.cc

```diff
@@ -286,5 +293,41 @@ HttpServerPropertiesCacheImpl::getOrCreateHttp3StatusTracker(const Origin& origi
   return *it->second.h3_status_tracker;
 }
 
+void HttpServerPropertiesCacheImpl::addCanonicalSuffix(std::string suffix) {
+  canonical_suffixes_.push_back(suffix);
+}
+
+absl::string_view HttpServerPropertiesCacheImpl::getCanonicalSuffix(absl::string_view hostname) {
+  for (const std::string& suffix : canonical_suffixes_) {
+    if (absl::EndsWith(hostname, suffix)) {
+      return suffix;
+    }
+  }
+  return "";
+}
+
+absl::optional<HttpServerPropertiesCache::Origin>
+HttpServerPropertiesCacheImpl::getCanonicalOrigin(absl::string_view hostname) {
+  absl::string_view suffix = getCanonicalSuffix(hostname);
+  if (suffix.empty()) {
+    return {};
+  }
+
+  auto it = canonical_alt_svc_map_.find(std::string(suffix));
+  if (it == canonical_alt_svc_map_.end()) {
+    return {};
+  }
+  return it->second;
+}
+
+void HttpServerPropertiesCacheImpl::maybeSetCanonicalOrigin(const Origin& origin) {
+  absl::string_view suffix = getCanonicalSuffix(origin.hostname_);
+  if (suffix.empty()) {
+    return;
+  }
+
+  canonical_alt_svc_map_[std::string(suffix)] = origin;
```

Comment:
Oh good point. I wasn't really think about the KV store other than to note that we don't need to persist the suffixes. But yeah, on load it really should just do the right thing.

Of course, because I wasn't thinking about it, I didn't write a test for it. And because there was no test for the right behavior, it had the wrong behavior :) This required passing in the suffixes to the constructor which I've done, and I've added a test. Thanks!